### PR TITLE
GpuCast.hasSideEffects now checks to see if the child expression has side-effects

### PR DESCRIPTION
**Pull request by andygrove**
Signed-off-by: Andy Grove <andygrove@nvidia.com>

Closes https://github.com/NVIDIA/spark-rapids/issues/5125

When expressions with side-effects were wrapped in a CAST then we treated them as though they did not have side-effects because the `GpuCast.hasSideEffects` method was not checking whether the child expression had side-effects.

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala [hunk 1]
```diff
@@ -1527,8 +1527,8 @@ case class GpuCast(
   import GpuCast._
 
   // when ansi mode is enabled, some cast expressions can throw exceptions on invalid inputs
-  override def hasSideEffects: Boolean = {
-    (child.dataType, dataType) match {
+  override def hasSideEffects: Boolean = super.hasSideEffects || {
+    (child. dataType, dataType) match {
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala:1531]: Nit: 
```suggestion
    (child.dataType, dataType) match {
```

**Review by jlowe — APPROVED** [review]
Minor whitespace nit otherwise lgtm.

**Review by jlowe — APPROVED** [review]

**Comment by jlowe** [comment]
build

**Comment by andygrove** [comment]
build

**Review by jlowe — APPROVED** [review]

**Review by jlowe — APPROVED** [review]

**Comment by jlowe** [comment]
build

**Pull request merged**